src/language/worker.ts

```diff
@@ -92,98 +96,85 @@ const executeAll = async (
   }
 };
 
-const findAndParse = async (
-  serviceProvider: CliServiceProvider,
-  databaseName: string,
-  collectionName: string
-) => {
-  const documents = await serviceProvider
-    .find(databaseName, collectionName, {}, { limit: 1 })
-    .toArray();
-
-  if (documents.length === 0) {
-    return [];
-  }
-
-  try {
-    const schema = await parseSchema(documents);
-
-    if (!schema || !schema.fields) {
-      return [];
-    }
-
-    const fields = schema.fields.map((item) => ({
-      label: item.name,
-      kind: CompletionItemKind.Field,
-    }));
-
-    return fields;
-  } catch (parseError) {
-    return [];
-  }
-};
-
+/**
+ * Get field names for autocomplete.
+ * @public
+ */
 const getFieldsFromSchema = async (
   connectionString: string,
-  connectionOptions: any,
+  connectionOptions: ConnectionOptions,
   databaseName: string,
   collectionName: string
-): Promise<any> => {
-  try {
-    const serviceProvider: CliServiceProvider =
-      await CliServiceProvider.connect(connectionString, connectionOptions);
+): Promise<[unknown, string[]?]> => {
+  let serviceProvider: CliServiceProvider;
+  let result: string[] = [];
 
-    const result = await findAndParse(
-      serviceProvider,
-      databaseName,
-      collectionName
+  try {
+    serviceProvider = await CliServiceProvider.connect(
+      connectionString,
+      connectionOptions
     );
-
-    return [null, result];
   } catch (error) {
     return [error];
   }
-};
 
-const prepareCompletionItems = (result: Document) => {
-  if (!result) {
-    return [];
+  try {
+    const documents = await serviceProvider
+      .find(databaseName, collectionName, {}, { limit: 1 })
+      .toArray();
+
+    if (documents.length) {
+      const schema = await parseSchema(documents);
+      if (schema?.fields) {
+        result = schema.fields.map((item) => item.name);
+      }
+    }
+  } catch (error) {
+    /* Squelch find and parse documents error */
   }
 
-  return result.databases.map((item) => ({
-    label: item.name,
-    kind: CompletionItemKind.Value,
-  }));
+  return [null, result];
 };
 
+/**
+ * Get database names for autocomplete.
+ * @public
+ */
 const getListDatabases = async (
   connectionString: string,
-  connectionOptions: any
-) => {
+  connectionOptions: ConnectionOptions
+): Promise<[unknown, Document[]?]> => {
   try {
-    const serviceProvider: CliServiceProvider =
-      await CliServiceProvider.connect(connectionString, connectionOptions);
+    const serviceProvider = await CliServiceProvider.connect(
+      connectionString,
+      connectionOptions
+    );
```

Comment:
I know this is a draft PR and pre-existing code, but I’m a bit surprised here … we are connecting a new `CliServiceProvider` for every autocompletion request? And never (explicitly) closing them? Those seem like bugs to me, might be good to fix that here and/or create follow-up tickets

Reply:
Yeah, this is a bug. Pushed changes, let me know if this is the right way.

Reply:
It answers the second part here, yeah :+1:

Reply:
I will probably refactor it in the current pr (not a separate one) to not create it each time. Makes total sense to change it.